akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSpec.scala

```diff
@@ -447,42 +484,330 @@ class InterpreterSpec extends InterpreterSpecKit {
+
+      upstream.onComplete()
+      lastEvents() should be(Set(OnNext(Vector(13, 14)), OnComplete))
+    }
+
+    "restart when onPush throws" in {
+      val stage = new RestartTestStage {
+        override def onPush(elem: Int, ctx: Context[Int]): Directive = {
+          if (elem <= 0) throw TE
+          else super.onPush(elem, ctx)
+        }
+      }
+
+      new TestSetup(Seq(
+        Map((x: Int) ⇒ x + 1, restartingDecider),
+        stage,
+        Map((x: Int) ⇒ x + 100, restartingDecider))) {
+
+        lastEvents() should be(Set.empty)
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(2)
+        lastEvents() should be(Set(OnNext(103)))
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(-1) // boom
+        lastEvents() should be(Set(RequestOne))
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(3)
+        lastEvents() should be(Set(OnNext(104)))
+      }
+    }
+
+    "restart when onPush throws after ctx.push" in {
+      val stage = new RestartTestStage {
+        override def onPush(elem: Int, ctx: Context[Int]): Directive = {
+          val ret = ctx.push(sum)
+          super.onPush(elem, ctx)
+          if (elem <= 0) throw TE
+          ret
+        }
+      }
+
+      new TestSetup(Seq(
+        Map((x: Int) ⇒ x + 1, restartingDecider),
+        stage,
+        Map((x: Int) ⇒ x + 100, restartingDecider))) {
+
+        lastEvents() should be(Set.empty)
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(2)
+        lastEvents() should be(Set(OnNext(103)))
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(-1) // boom
+        lastEvents() should be(Set(RequestOne))
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(3)
+        lastEvents() should be(Set(OnNext(104)))
+      }
+    }
+
+    "restart when onPull throws" in {
+      val stage = new RestartTestStage {
+        override def onPull(ctx: Context[Int]): Directive = {
+          if (sum < 0) throw TE
+          super.onPull(ctx)
+        }
+      }
+
+      new TestSetup(Seq(
+        Map((x: Int) ⇒ x + 1, restartingDecider),
+        stage,
+        Map((x: Int) ⇒ x + 100, restartingDecider))) {
+
+        lastEvents() should be(Set.empty)
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(2)
+        lastEvents() should be(Set(OnNext(103)))
+
+        downstream.requestOne()
+        lastEvents() should be(Set(RequestOne))
+        upstream.onNext(-5) // this will trigger failure of next requestOne (pull)
+        lastEvents() should be(Set(OnNext(99)))
+
+        downstream.requestOne() // this failed, but resume will pull
```

Comment:
Comment is wrong, this is "restart" case. Will see how this will work generally for example if that onPull happened after absorbTermination.